### PR TITLE
Remove x87 floating point support

### PR DESCRIPTION
**Pull request by BradleyWood**
Refactoring FPTreeEvaluator in process of removing x87 support. Removed conditional logic to always use SSE.

Refactor float/double to long conversion on x86 to use SSE instead of x87 instructions.

Issue: #946

### CODE REVIEW

**Comment by fjeremic** [comment]
Looks like you have duplicate commits here and a missing line break in the first one.

**Comment by BradleyWood** [comment]
I will squash those two before I finish with this.

**Review by 0xdaryl — COMMENTED** [review]
In general the code changes here look good. As expected, there is some trickiness on the f/d -> i/l paths, as well as floating point remainder that we need to sort out.

The existing conversion operations adhere to Java semantics, in particular on the overflow, infinity, and NaN cases.  OMR itself doesn't define default floating point semantics (it probably should, but that is something to come later).  Because of some of the complications introduced to ensure the converted value follows Java semantics (e.g., the use of snippets and runtime assembler helpers) I think it makes sense to relocate the `f2i`/`f2l`/`d2i`/`d2l` evaluators to OpenJ9 (and taking the snippets with them if that is their only use).  In its place, OMR could implement a much simpler implementation more akin to C conversion semantics (using round-to-zero, i.e., truncation) where handling of the special values is not done.

On 64-bit, this means using the SSE conversion instructions without checking for the indefinite result (because handling the special result would mean we are going to impose floating point semantics that we don't have defined in OMR.  On 32-bit, this means that `FISTP m32/m64` instructions can be used if the FPCW is set to truncation, or on processors that support SSE3 use the `FISTTP m32/m64` instructions which do truncation without having to set the control word.  You will have to add support for `FISTTP` and check when we are allowed to use it.  This can be used in OMR without requiring support for x87 register assignment by producing pre-encoded instructions.  That is,
```
   mov [m32], xmm6
   fld m32                ; Loads m32 to ST0 always; encoding doesn't require x87 RA
   fisttp intm32          ; Converts, truncates, pops x87 stack, and stores in intm32.
                          ; Encoding doesn't require x87 RA.
```
If you use `FISTP` on older processors without SSE3 support you will have to preserve the `FPCW`, set the `FPCW` to truncating rounding mode, do the `FISTP`, and then reset the `FPCW` back to what it was before.

Floating point remainder is trickier.  At the moment, the FP modulus implementation on 64-bit won't actually work because it requires a helper to do the computation.  However, the helper implementation doesn't exist in OMR so non-Java runtimes will simply crash.  I think the floating point remainder evaluators should be relocated to OpenJ9.  In OMR, I think we should leave these opcodes as unimplemented on x86.  Like the conversion opcodes, the semantics of floating point modulus in OMR is unspecified, so providing an implementation in OMR is arbitrary anyway.

**Comment by fjeremic** [comment]
@BradleyWood there are some conflicts here.

**Comment by BradleyWood** [comment]
[Fixing Conflicts](https://github.com/eclipse/omr/compare/41e3cea913f87bfa6a69fcd392e821c67e90171e..770a073c7ba2356a730462062c23b3e214ad5d73)

**Comment by 0xdaryl** [comment]
Jenkins build xlinux,win,osx

File: compiler/x/codegen/FPTreeEvaluator.cpp [hunk 1]
```diff
@@ -515,7 +437,10 @@ TR::Register *OMR::X86::TreeEvaluator::fpReturnEvaluator(TR::Node *node, TR::Cod
 
    const TR::X86LinkageProperties &linkageProperties = cg->getProperties();
    TR::RealRegister::RegNum machineReturnRegister =
-      (returnRegister->isSinglePrecision())? linkageProperties.getFloatReturnRegister() : linkageProperties.getDoubleReturnRegister();
+         (returnRegister->isSinglePrecision())? linkageProperties.getFloatReturnRegister() : linkageProperties.getDoubleReturnRegister();
+
+   TR_ReturnInfo returnInfo = (returnRegister->isSinglePrecision()) ? TR_FloatXMMReturn : TR_DoubleXMMReturn;
```
Comment (0xdaryl) [inline compiler/x/codegen/FPTreeEvaluator.cpp:442]: Can you sink this down to where it is actually used?  It makes the code more readable.

File: compiler/x/codegen/FPTreeEvaluator.cpp [hunk 2]
```diff
@@ -789,29 +701,21 @@ TR::Register *OMR::X86::TreeEvaluator::fpRemEvaluator(TR::Node *node, TR::CodeGe
...
+      {
+      // TODO: We should do this for IA32 eventually
+      TR::SymbolReference *helperSymRef = cg->symRefTab()->findOrCreateRuntimeHelper(nodeIsDouble ? TR_AMD64doubleRemainder : TR_AMD64floatRemainder);
+      targetRegister = TR::TreeEvaluator::performHelperCall(node, helperSymRef, nodeIsDouble ? TR::dcall : TR::fcall, false, cg);
       }
    else
       {
-      targetRegister = TR::TreeEvaluator::commonFPRemEvaluator(node, cg, nodeIsDouble);
+      TR::SymbolReference *helperSymRef = cg->symRefTab()->findOrCreateRuntimeHelper(nodeIsDouble ? TR_IA32doubleRemainderSSE : TR_IA32floatRemainderSSE);
+      targetRegister = TR::TreeEvaluator::performHelperCall(node, helperSymRef, nodeIsDouble ? TR::dcall : TR::fcall, false, cg);
```
Comment (0xdaryl) [inline compiler/x/codegen/FPTreeEvaluator.cpp:716]: To save code footprint can you rewrite this as:
```
TR_RuntimeHelper remainderHelper;
if (cg->comp()->target().is64Bit())
   {
   remainderHelper = nodeIsDouble ? TR_AMD64doubleRemainder : TR_AMD64floatRemainder;
   }
else
   {
   remainderHelper = nodeIsDouble ? TR_IA32doubleRemainderSSE : TR_IA32floatRemainderSSE;
   }

TR::SymbolReference *helperSymRef = cg->symRefTab()->findOrCreateRuntimeHelper(remainderHelper);
targetRegister = TR::TreeEvaluator::performHelperCall(node, helperSymRef, nodeIsDouble ? TR::dcall : TR::fcall, false, cg);

```

File: compiler/x/codegen/FPTreeEvaluator.cpp [hunk 3]
```diff
@@ -1328,131 +967,107 @@ TR::Register *OMR::X86::TreeEvaluator::fpConvertToLong(TR::Node *node, TR::Symbo
...
+      // We can't compare with 0x8000000000000000.
+      // Instead, rotate left 1 bit and compare with 0x0000000000000001.
+      generateRegInstruction(TR::InstOpCode::ROL8Reg1, node, targetRegister, cg);
+      generateRegImmInstruction(TR::InstOpCode::CMP8RegImms, node, targetRegister, 1, cg);
+      generateLabelInstruction(TR::InstOpCode::JE4, node, exceptionLabel, cg);
+      }
+   else
+      {
+      generateRegImmInstruction(TR::InstOpCode::CMP4RegImm4, node, targetRegister, INT_MIN, cg);
+      generateLabelInstruction(TR::InstOpCode::JE4, node, exceptionLabel, cg);
```
Comment (0xdaryl) [inline compiler/x/codegen/FPTreeEvaluator.cpp:1063]: Put this JE4 (and the one in the if branch) after the if/else paths to reduce code footprint.

File: compiler/x/codegen/FPTreeEvaluator.cpp [hunk 4]
```diff
@@ -1880,54 +1375,12 @@ TR::Register *OMR::X86::TreeEvaluator::fRegStoreEvaluator(TR::Node *node, TR::Co
...
-                     cg->stopUsingRegister(childGlobalReg);
-                     cg->machine()->getFPStackRegisterNode(fpStackSlot)->setRegister(popReg);
-                     break;
-                     }
-                  }
-               }
-            }
-         }
+   if (globalReg->getKind() != TR_FPR)
+      globalReg = TR::TreeEvaluator::coerceFPRToXMMR(child, globalReg, cg);
```
Comment (0xdaryl) [inline compiler/x/codegen/FPTreeEvaluator.cpp:1379]: Delete these two lines.  The globalReg can't be anything other than a TR_FPR.  You could add a TR_FATAL_ASSERT for this instead.

File: compiler/x/codegen/FPTreeEvaluator.cpp [hunk 5]
```diff
@@ -1939,58 +1392,12 @@ TR::Register *OMR::X86::TreeEvaluator::dRegStoreEvaluator(TR::Node *node, TR::Co
...
-
-      machine->setXMMGlobalRegister(globalRegNum - machine->getNumGlobalGPRs(), globalReg);
-      cg->decReferenceCount(child);
-      return globalReg;
-      }
-   else
-      {
-      machine->setFPStackRegister(globalRegNum - machine->getNumGlobalGPRs(), toX86FPStackRegister(globalReg));
+   if (globalReg->getKind() != TR_FPR)
+      globalReg = TR::TreeEvaluator::coerceFPRToXMMR(child, globalReg, cg);
```
Comment (0xdaryl) [inline compiler/x/codegen/FPTreeEvaluator.cpp:1396]: Delete these two lines. The globalReg can't be anything other than a TR_FPR. You could add a TR_FATAL_ASSERT for this instead.

**Review by 0xdaryl — CHANGES_REQUESTED** [review]
I did a more thorough final review of this and have a few final comments.

**Comment by BradleyWood** [comment]
See [force-push](https://github.com/eclipse/omr/compare/83e0c240b44271fbf5c2cf5b524f2a36a68160cc..3f3242b2fcc50a09c5163540ef40a0e7860290e3)

**Comment by 0xdaryl** [comment]
Jenkins build xlinux,win,osx

**Review by 0xdaryl — APPROVED** [review]

**Comment by 0xdaryl** [comment]
Can you rebase to resolve the merge conflicts please?

**Comment by BradleyWood** [comment]
> Can you rebase to resolve the merge conflicts please?

There was a single conflict inside a function that was deletec. Should not affect the build.

**Comment by fjeremic** [comment]
Jenkins build all

**Pull request merged**